### PR TITLE
Debian version of the Installing Foreman guide

### PR DESCRIPTION
**Pull request by melcorr**
This is the early draft of the Foreman installation guide on Debian. 
If @lzap thinks that this could work, I will implement the changes for external Smart Proxy on Debian also. 
As part of 

Bug 1803033 - [upstream] add Installing Foreman on Debian system guide.
https://bugzilla.redhat.com/show_bug.cgi?id=1803033

### CODE REVIEW

File: guides/common/attributes.adoc [hunk 1]
```diff
@@ -24,6 +24,9 @@ ifeval::["{build}" == "foreman"]
 :foreman-example-com: foreman.example.com
 :installer-scenario: foreman-installer --scenario katello
 :installer-scenario-smartproxy: foreman-installer --no-enable-foreman
+:install: yum install
+:clean: yum clean
+:remove: yum remove
```
Comment (lzap) [inline guides/common/attributes.adoc]: I'd probably name these `package-install`, those look too generic to me.

File: guides/common/attributes.adoc [hunk 2]
```diff
@@ -54,6 +57,9 @@ ifeval::["{build}" == "satellite"]
 :foreman-installer: satellite-installer
 :installer-scenario: satellite-installer --scenario satellite
 :installer-scenario-smartproxy: satellite-installer --scenario capsule
+:install: yum install
+:clean: yum clean
+:remove: yum remove
```
Comment (lzap) [inline guides/common/attributes.adoc]: Isn't this supposed to be `foreman-maintain` now?

Comment (melcorr) [reply guides/common/attributes.adoc]: @lzap Capsule doesn't have satellite-maintain so we need yum some of the time!

File: guides/common/modules/ports_prerequisites.adoc [hunk 3]
```diff
@@ -9,7 +9,13 @@ The following tables indicate the destination port and the direction of network
 | 443 | TCP | HTTPS | Subscription Management Services (access.redhat.com) and connecting to the Red{nbsp}Hat CDN (cdn.redhat.com).
 |====
 
+ifeval::["{Build}" == "satellite"]
 Except in the case of a disconnected {Project}, {ProjectServer} needs access to the Red{nbsp}Hat CDN. For a list of IP addresses used by the Red{nbsp}Hat CDN (cdn.redhat.com), see the Knowledgebase article https://access.redhat.com/articles/1525183[Public CIDR Lists for Red Hat] on the Red{nbsp}Hat Customer Portal.
+endif::[]
+
+ifeval::["{Build}" == "foreman"]
+If you plan to use Red{nbsp}Hat services, {ProjectServer} needs access to the Red{nbsp}Hat CDN. For a list of IP addresses used by the Red{nbsp}Hat CDN (cdn.redhat.com), see the Knowledgebase article https://access.redhat.com/articles/1525183[Public CIDR Lists for Red Hat] on the Red{nbsp}Hat Customer Portal.
+endif::[]
```
Comment (lzap) [inline guides/common/modules/ports_prerequisites.adoc]: Here is a problem, we will have many (if not all) blocks which should be rendered for both "foreman" and "foreman-deb". Unfortunately, I've checked the asciidoctor codebase and the "ifeval" statement does not provide regular matching or globbing (like "foreman*"). Also there is no "else" keyword available. However this syntax does work:

ifeval::["{Build}" == "satellite"]
Downstream only thing.
endif::[]
ifeval::["{Build}" != "satellite"]
Upstream only thing for both CentOS and Debian.
endif::[]

File: guides/common/modules/ref_capsule-ports-and-firewalls-requirements.adoc [hunk 4]
```diff
@@ -36,6 +36,16 @@ Qpid dispatch router
 |8000 |TCP |HTTPS |Operating System installers like Anaconda to download kickstart templates to hosts,
 and for downloading iPXE firmware
 endif::[]
+ifeval::["{Build}" == "foreman-deb"]
+|Port |Protocol |Service |Required for
+|80 |TCP |HTTP |Operating System installers like Anaconda, yum, and, if you use the Katello plug-in, for obtaining Katello certificate
+updates
```
Comment (lzap) [inline guides/common/modules/ref_capsule-ports-and-firewalls-requirements.adoc]: Is this some broken line?

File: guides/common/modules/ref_supported-operating-systems.adoc [hunk 5]
```diff
@@ -67,3 +66,39 @@ endif::[]
...
+The following operating systems are supported by `{foreman-installer}`, have packages, and are tested for deploying {Project}:
+
+.Operating Systems supported by {foreman-installer}
+|====
+| Operating System | Architecture | Notes
+
+| Ubuntu 16.04 (Xenial) | i386, amd64, aarch64 |
+| Ubuntu 18.04 (Bionic) | i386, amd64, aarch64 |
+| Debian 9 (Stretch) | i386, amd64, aarch64 |
+|====
```
Comment (lzap) [inline guides/common/modules/ref_supported-operating-systems.adoc]: We only support `amd64` on all OSes.

File: guides/common/modules/ref_supported-operating-systems.adoc [hunk 6]
```diff
@@ -67,3 +66,39 @@ endif::[]
...
+*  Fedora
+*  Ubuntu
+*  Debian
+*  OpenSUSE
+*  SLES
+*  CoreOS
+*  Solaris
+*  FreeBSD
+*  Juniper Junos
+*  Cisco NX-OS
```
Comment (lzap) [inline guides/common/modules/ref_supported-operating-systems.adoc]: I'd ditch this list of "known to work". Zero information, mostly misleading.

File: guides/common/modules/ref_system-requirements.adoc [hunk 7]
```diff
@@ -5,7 +5,7 @@
 The following requirements apply to the networked base operating system:
 
 * x86_64 architecture
-ifeval::["{context}" == "satellite"]
+ifeval::["{Build}" == "satellite"]
```
Comment (lzap) [inline guides/common/modules/ref_system-requirements.adoc]: Honestly, "Build" already drives me crazy, I'd find "build" more natural. How about either changing that or introducing an alias: `:build: {Build}` in attributes? :-)

File: guides/doc-Installing_Foreman_on_Debian/master.adoc [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+include::common/modules/AWS_Installing_Satellite_AWS.adoc[leveloffset=+1]
+
+// Installing {SmartProxy} on AWS
+include::common/modules/AWS_Installing_Capsule_on_AWS.adoc[leveloffset=+1]
+
+
+:numbered!:
+
+// Applying Custom Configuration to {Project}
+include::common/modules/applying_custom_configuration_to_red_hat_satellite.adoc[leveloffset=+1]
```
Comment (lzap) [inline guides/doc-Installing_Foreman_on_Debian/master.adoc]: Modularization kinda works here, hmmmm.

File: guides/common/attributes.adoc (hunk 1 above)
Comment (lzap) [reply guides/common/attributes.adoc]: Also clean can have different meaning for other package managers, I'd probably not bother in that particular case.

**Comment by lzap** [comment]
Ok looks nice, I have one major comment tho. Introducing of "foreman-deb" actually breaks many of the existing ifevals we currently have. We need to negate this to `ifeval::["{Build}" != "satellite"]` I believe in all cases:

```
[lzap@box guides]$ ag ifeval | grep foreman
doc-Provisioning_Guide/master.adoc:4:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Introduction.adoc:44:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-GCE.adoc:111:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:86:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:98:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:154:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:279:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:292:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Networking.adoc:382:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-OpenStack.adoc:132:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-VMWare.adoc:204:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-VMWare.adoc:213:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-VMWare.adoc:298:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-VMWare.adoc:333:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-Libvirt.adoc:13:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-Libvirt.adoc:263:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/app-init_script.adoc:5:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:239:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc:261:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-Kubevirt.adoc:33:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Common_Compute_Resource_Prereqs.adoc:3:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Common_Compute_Resource_Prereqs.adoc:12:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:18:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:94:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:107:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:139:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:227:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:399:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Bare_Metal.adoc:425:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:14:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:24:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:34:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:54:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:168:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:465:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:478:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Resources.adoc:486:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc:141:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc:238:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc:251:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc:272:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc:280:ifeval::["{Build}" == "foreman"]
doc-Installing_Satellite_Server_Connected/master.adoc:4:ifeval::["{Build}" == "foreman"]
doc-Installing_Capsule_Server/master.adoc:4:ifeval::["{Build}" == "foreman"]
common/modules/snip_firewalld.adoc:1:ifeval::["{Build}" == "foreman"]
common/modules/ref_capsule-ports-and-firewalls-requirements.adoc:12:ifeval::["{Build}" == "foreman"]
common/modules/ref_capsule-ports-and-firewalls-requirements.adoc:29:ifeval::["{Build}" == "foreman"]
common/modules/synchronizing_the_satellite_tools_repository.adoc:4:ifeval::["{Build}" == "foreman"]
common/modules/AWS_Deployment_Scenarios.adoc:12:ifeval::["{Build}" == "foreman"]
common/modules/proc_adding-life-cycle-environments.adoc:5:ifeval::["{Build}" == "foreman"]
common/modules/using_external_databases.adoc:4:ifeval::["{Build}" == "foreman"]
common/modules/installing_overview.adoc:9:ifeval::["{Build}" == "foreman"]
common/modules/ref_supported-operating-systems.adoc:9:ifeval::["{Build}" == "foreman"]
common/modules/proc_registering-to-satellite-server.adoc:5:ifeval::["{Build}" == "foreman"]
common/modules/proc_restricting-access-to-mongod.adoc:5:ifeval::["{Build}" == "foreman"]
common/modules/proc_attaching-satellite-infrastructure-subscription.adoc:5:ifeval::["{Build}" == "foreman"]
common/modules/proc_uninstalling-capsule-server.adoc:5:ifeval::["{Build}" == "foreman"]
common/modules/enabling_the_satellite_tools_repository.adoc:5:ifeval::["{Build}" == "foreman"]
common/assembly_installing-capsule-server.adoc:26:ifeval::["{Build}" == "foreman"]
common/attributes.adoc:21:ifeval::["{build}" == "foreman"]
doc-Installing_Satellite_Server_Disconnected/master.adoc:4:ifeval::["{Build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-RHV.adoc:7:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-RHV.adoc:60:ifeval::["{build}" == "foreman"]
doc-Provisioning_Guide/topics/Virt-RHV.adoc:236:ifeval::["{Build}" == "foreman"]
README.md:80:   ifeval::["{build}" == "foreman"]
```

Also explain this in the README for newcomers.

File: guides/common/modules/ports_prerequisites.adoc (hunk 3 above)
Comment (melcorr) [reply guides/common/modules/ports_prerequisites.adoc]: This, for me can just be absent from the Debian guide as they will never use Red Hat CDN for  their Foreman server will they? @lzap

File: guides/common/modules/ref_capsule-ports-and-firewalls-requirements.adoc (hunk 4 above)
Comment (melcorr) [reply guides/common/modules/ref_capsule-ports-and-firewalls-requirements.adoc]: @lzap this renders fine! believe it or not!

File: guides/common/modules/ref_system-requirements.adoc (hunk 7 above)
Comment (melcorr) [reply guides/common/modules/ref_system-requirements.adoc]: This Build is a sneaky thing that just bled in. I'll s/Build/build throughout.

**Comment by melcorr** [comment]
This comment went completely outside of the scope of the Installation Guide for Debian, but at the same time, it is good now that we can build a Provisioning Guide for Debian also and we can continue to think about each ifeval in this context so if we need to generate these guides, we can do so easily .

I went through every one of these and here are my findings

> doc-Provisioning_Guide/master.adoc:4:ifeval::["{Build}" == "foreman"]

Agree. Changed to != satellite

> doc-Provisioning_Guide/topics/Introduction.adoc

This references Katello plug-in, so should be foreman ifeval only

> doc-Provisioning_Guide/topics/Cloud-GCE.adoc

foreman ifeval references the Katello plug-in, not applicable to Debian, so must remain. 

> doc-Provisioning_Guide/topics/Networking.adoc

Changed 5 to != kept one Katello reference

> doc-Provisioning_Guide/topics/Cloud-OpenStack.adoc

This is a Katello reference and must be kept to foreman ifeval

> doc-Provisioning_Guide/topics/Virt-VMWare.adoc

Two remain as Katello references, 

There is one for configuring cloud-init that will need a foreman-deb ifeval, once someone can help us figure out the steps. 

> doc-Provisioning_Guide/topics/Virt-Libvirt.adoc

Replaced one ifeval and kept one foreman ifeval for referencing Katello plugin.

> doc-Provisioning_Guide/topics/Common_Compute_Resource_Prereqs.adoc

Created a foreman-deb ifeval for Instalation media without any reference to Katello plugin

> doc-Provisioning_Guide/topics/Bare_Metal.adoc

Made two updates. Others relate to Katello plugin

> doc-Provisioning_Guide/topics/Resources.adoc

Added ifeval::["{build}" != "foreman-deb"] x 2 for stuff that does not apply to foreman-deb scenarios but may apply to Satellite and Foreman

Replaced all except those that reference Katello. 

> doc-Provisioning_Guide/topics/Cloud-Amazon_EC2.adoc

Replaced all but Katello references

> doc-Installing_Satellite_Server_Connected/master.adoc
> doc-Installing_Capsule_Server/master.adoc

these are irrelevant as Debian has its own directory so we can manipulate the master.adoc with only the necessary modules, but it is also no harm and might be future-proofing other versions of this guide for use in the community. 

> common/modules/snip_firewalld.adoc

This change makes sense

> common/modules/ref_capsule-ports-and-firewalls-requirements.adoc

For now, I have added a foreman-deb ifeval to say for Debian deployments ignore all references in the table to Katello plug-in, maybe there is a more elegant way to handle this. 


> common/modules/synchronizing_the_satellite_tools_repository.adoc
> common/modules/proc_adding-life-cycle-environments.adoc
> common/modules/using_external_databases.adoc

These are not included in the Debian master.adoc, so no need to edit. 

> common/modules/AWS_Deployment_Scenarios.adoc

Updated the ifeval for all non-Satellite scenarios

> common/modules/installing_overview.adoc

Added ifeval for Debian scenarios.

> common/modules/ref_supported-operating-systems.adoc

have replaced the foreman ifeval with !=

> common/modules/proc_registering-to-satellite-server.adoc

This is not in the master.adoc for Debian, so no changes needed. 

> common/modules/proc_restricting-access-to-mongod.adoc
> common/modules/proc_attaching-satellite-infrastructure-subscription.adoc
> common/modules/proc_uninstalling-capsule-server.adoc
> common/modules/enabling_the_satellite_tools_repository.adoc
> common/assembly_installing-capsule-server.adoc
> doc-Installing_Satellite_Server_Disconnected/master.adoc

Applies only to Katello scenarios, or Satellite only, and not included in master.adoc for Debian. 

> common/attributes.adoc

Not going to touch this for the moment. Might revise when this is finished. 

> doc-Provisioning_Guide/topics/Virt-RHV.adoc

Replaced one but the other two relate to Katello, so are "foreman"

Head fried now :-)

File: guides/common/attributes.adoc (hunk 1 above)
Comment (melcorr) [reply guides/common/attributes.adoc]: Let's keep this and ask the community members working with Debian to let us know if an alternative command would work for them.

**Comment by melcorr** [comment]
Hi @Baboon92 

This is the WIP PR for the Debian guide. 
I will tag you in different comments here, and please feel free to respond when you have the time

If you want to preview, please follow these instructions to install the packages so you can preview locally: 

https://github.com/theforeman/foreman-documentation/blob/master/guides/README.md

If you need help with anything, please feel free to contact me.

**Comment by melcorr** [comment]
@lzap all the latest changes from master are merged with this branch.

**Comment by lzap** [comment]
Mel, I see you are merging back into this branch. That's never a good idea, I mean it has its own use but in this case what you want to do is to rebase this PR on top of master:

```
git checkout master
git pull
git checkout debian-experiment
git rebase master
git mergetool # or resolve conflicts with your preferred tool
git push
```

**Comment by melcorr** [comment]
@lzap  thanks. I didn't know this. I will do this in future. I thought that was a good method to find and fix conflicts, but won't do it again.

**Comment by melcorr** [comment]
Hi @lzap I followed the rebase steps.
Thanks!

**Comment by melcorr** [comment]
Rebased with changes from latest merges.

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 9]
```diff
@@ -234,8 +234,9 @@ You must perform this procedure on all {SmartProxies}.
 
 . Install the `ipxe-bootimgs` RPM package:
 +
+[options="nowrap" subs="+quotes,attributes"]
 ----
-# yum install ipxe-bootimgs
+# {install} ipxe-bootimgs
```
Comment (lzap) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc]: This should be `{package-install}`.

File: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc [hunk 10]
```diff
@@ -56,7 +56,7 @@ Before you begin, ensure that the following conditions are met:
 [options="nowrap" subs="+quotes,attributes"]
 ifeval::["{build}" == "satellite"]
 ----
-# yum install foreman-discovery-image
+# {install} foreman-discovery-image
```
Comment (lzap) [inline guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc]: This should be `{package-install}`.

File: guides/doc-Provisioning_Guide/topics/app_create_images.adoc [hunk 11]
```diff
@@ -490,10 +496,11 @@ NOZEROCONF=yes
 +
 . Un-register the virtual machine so that the resulting image does not contain the same subscription details for every instance cloned based on it:
 +
+[options="nowrap" subs="+quotes,attributes"]
 ----------
 # subscription-manager repos --disable=*
 # subscription-manager unregister
-# yum clean all
+# {clean} all
```
Comment (lzap) [inline guides/doc-Provisioning_Guide/topics/app_create_images.adoc]: This should be `{package-clean}`.

File: guides/common/modules/proc_configuring-repositories.adoc [hunk 12]
```diff
@@ -43,11 +43,11 @@ endif::[]
...
 
 +
-. Clear any `yum` metadata:
+. Clear any metadata:
 +
-[options="nowrap"]
+[options="nowrap" subs="+quotes,attributes"]
 ----
-# yum clean all
+# {clean} all
```
Comment (lzap) [inline guides/common/modules/proc_configuring-repositories.adoc]: This should be `{package-clean}`.

File: guides/doc-Provisioning_Guide/topics/Virt-VMWare.adoc [hunk 13]
```diff
@@ -346,11 +346,12 @@ endif::[]
 ----
 . Use the following commands to clean the image:
 +
+[options="nowrap" subs="+quotes,attributes"]
 ----
 # systemctl stop rsyslog
 # systemctl stop auditd
 # package-cleanup -oldkernels -count=1
-# yum clean all
+# {clean} all
```
Comment (lzap) [inline guides/doc-Provisioning_Guide/topics/Virt-VMWare.adoc]: This should be `{package-clean}`.

**Comment by melcorr** [comment]
lol thanks for the review @lzap - this was supposed to focus on Installation guide but we may as well update the attributes everywhere as part of this. Working on this now.

**Comment by lzap** [comment]
Thanks, squashed and merged

**Pull request merged**